### PR TITLE
GetBatchById-SecondMerge

### PR DESCRIPTION
**Pull request by tavisca-adsouza**

### CODE REVIEW

File: src/Tavisca.TMS.Core/Services/BatchService.cs [hunk 1]
```diff
@@ -9,16 +9,19 @@
...
 using System.Threading.Tasks;
+using Tavisca.TMS.Contracts.Models.BatchModels;
 
 namespace Tavisca.TMS.Core
 {
     public class BatchService : IBatchService
     {
         private readonly IBatchRepository _batchRepository;
         private readonly IValidator<Batch> _validator;
+        private readonly GuidIdValidator _idValidator;
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Core/Services/BatchService.cs]: Why we need GuidIdValidator? Isn't a single validator sufficient to handle the validations?
Also, it is not a good practice to create validators by technical names like guid or string. Create validators by business names.

File: src/Tavisca.TMS.Core/Services/BatchService.cs [hunk 2]
```diff
@@ -154,16 +157,24 @@ public async Task<ApiResponse> Update(string trainingId, Batch batch)
...
         public async Task<ApiResponse> GetBatchById(string batchId)
         {
-
-
-            BatchResponse<BatchDetails> batchDetails = await _batchRepository.GetBatchById(batchId);
-            if (batchDetails == null || batchDetails.data == null)
+            try
             {
-                return new ApiFailureResponse(batchDetails.ErrorMessage, batchDetails.ErrorCode);
+                //if (_idValidator.IsIdValidGuid(batchId) == false)
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Core/Services/BatchService.cs]: Remove commented code

File: src/Tavisca.TMS.Core/Validators/GuidIdValidator.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+namespace Tavisca.TMS.Core
+{
+    public class GuidIdValidator : IGuidIdValidator
+    {
+        
+
+        public bool IsIdValidGuid(string id)
+        {
+            Guid guidOutput;
+            bool isValid = Guid.TryParse(id, out guidOutput);
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Core/Validators/GuidIdValidator.cs]: Avoid creation of unnecessary variables.

File: src/Tavisca.TMS.Persistence/Dal/BatchDal.cs [hunk 4]
```diff
@@ -225,64 +225,80 @@ private BatchDto Extract(MySqlDataReader reader)
...
-                        var reader = cmd.ExecuteReader();
-
-                        if (reader.HasRows == false)
+                        try
                         {
-                            throw new RowNotInTableException();
+                        
+                        batchResponse.data = await GetBatchInformation();
+
+                        async Task<BatchDto> GetBatchInformation()
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Persistence/Dal/BatchDal.cs]: What code is this? Whether its getting build?

File: src/Tavisca.TMS.Persistence/MySqlConnection.cs [hunk 5]
```diff
@@ -15,7 +15,7 @@ public MySqlConnectionProvider()
             var sqlPassword = Environment.GetEnvironmentVariable("sql_password", EnvironmentVariableTarget.Machine);
 
             var dbName = "tms_db";
-            var _connectionString = $"server=localhost;port=3306;Database={dbName}; uid={ sqlUsername };pwd= { sqlPassword }; convert zero datetime=True;";
+            var _connectionString = $"server=localhost;port=3306;Database={dbName}; uid={ sqlUsername };pwd= { sqlPassword }; convert zero datetime=True; ";
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Persistence/MySqlConnection.cs]: Its not a good practice to keep connStrings in a class. You must move them to app.config or web.config

File: src/Tavisca.TMS.Persistence/Repositories/BatchRepository.cs [hunk 6]
```diff
@@ -58,24 +59,32 @@ public async Task<List<BatchDetails>> GetByTrainingID(string trainingId)
...
+                    repoResponse.ErrorCode = dalResponse.ErrorCode;
+                }
+                return repoResponse;
             }
-            else
+            catch(Exception ex)
             {
-                batchResponse.ErrorMessage = dalResponse.ErrorMessage;
-                batchResponse.ErrorCode = dalResponse.ErrorCode;
+                throw ex;
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Persistence/Repositories/BatchRepository.cs]: What is a purpose of exception handling here?

File: src/Tavisca.TMS.Tests/Core/Services/BatchServiceFixture.cs [hunk 7]
```diff
@@ -124,15 +126,63 @@ public void Batch_service_should_call_repository_CreateBatch_method_once()
...
         {
-            string value = "1";
-            _batchRepository.Setup(x => x.GetBatchById(value)).ReturnsAsync(new BatchResponse<BatchDetails>());
+            string id = "1";
+            //_guidIdValidator.Setup(x => x.IsIdValidGuid(id)).Returns(true);
+            _batchRepository.Setup(x => x.GetBatchById(id)).ReturnsAsync(new BatchResponse<GetBatchByIdDetails>());
 
-            _service.GetBatchById(value);
-            _batchRepository.Verify(x => x.GetBatchById(value), Times.Once);
+            var t = await _service.GetBatchById(id);
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Tests/Core/Services/BatchServiceFixture.cs]: don't use variables like 't'

File: src/Tavisca.TMS.Tests/Core/Services/BatchServiceFixture.cs [hunk 8]
```diff
@@ -124,15 +126,63 @@ public void Batch_service_should_call_repository_CreateBatch_method_once()
...
+        public async Task Batch_service_should_rethrow_exception_to_the__method_that_called_it_when_any_part_of_the_method_throws_an_exception()
+        {
+            string id = "1";
+            _batchRepository.Setup(x => x.GetBatchById(id)).Throws(new NotSupportedException());
+            Exception ex = await Assert.ThrowsAsync<NotSupportedException>(() => _service.GetBatchById(id));
+        }
+
+        //[Fact]
+        //public async Task Batch_service_should_throw_error_if_IsIdValidGuid_throws_an_error()
+        //{
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Tests/Core/Services/BatchServiceFixture.cs]: remove commented code.

Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Tests/Core/Services/BatchServiceFixture.cs]: Done.

File: src/Tavisca.TMS.Tests/Core/Services/BatchServiceFixture.cs (hunk 7 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Tests/Core/Services/BatchServiceFixture.cs]: Done for my test cases.

File: src/Tavisca.TMS.Persistence/Repositories/BatchRepository.cs (hunk 6 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Persistence/Repositories/BatchRepository.cs]: To be handled in Controller layer as a BadRequest HTTP code response.

File: src/Tavisca.TMS.Persistence/MySqlConnection.cs (hunk 5 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Persistence/MySqlConnection.cs]: Will notify the person who wrote this.

File: src/Tavisca.TMS.Core/Validators/GuidIdValidator.cs (hunk 3 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Core/Validators/GuidIdValidator.cs]: This method was deleted.

File: src/Tavisca.TMS.Core/Services/BatchService.cs (hunk 2 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Core/Services/BatchService.cs]: Done.

File: src/Tavisca.TMS.Core/Services/BatchService.cs (hunk 1 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Core/Services/BatchService.cs]: Removed.

File: src/Tavisca.TMS.Persistence/Dal/BatchDal.cs (hunk 4 above)
Comment (tavisca-adsouza) [reply src/Tavisca.TMS.Persistence/Dal/BatchDal.cs]: This is actually a Task method that I included inside because of the using keyword scope issue

**Pull request merged**